File: C++/526.cpp
__________________________________________________________________________________________________
sample 4 ms submission
class Solution {
public:
    Solution() {

    }

    int countArrangement(int N) {
        SuitableNumbers.assign(16, vector<int>());
        for (int i = 1; i <= 15; ++i) {
            for (int j = 1; j <= 15; ++j) {
                if (i % j == 0 || j % i == 0) {
                    SuitableNumbers[i].push_back(j - 1);
                }
            }
        }
        memset(WasCalced, false, sizeof(WasCalced));
        return getCount((1 << N) - 1, N, N);
    }
private:
    unsigned long long getCount(int mask, int usedCount, int n) {
        if (usedCount == 0) {
            return 1;
        }
        if (WasCalced[mask]) {
            return CountByMask[mask];
        }
        WasCalced[mask] = true;
        CountByMask[mask] = 0;
        for (int i = 0; i < SuitableNumbers[usedCount].size() && SuitableNumbers[usedCount][i] < n; ++i) {
            if (mask & (1 << (SuitableNumbers[usedCount][i]))) {
                CountByMask[mask] += getCount(mask ^ (1 << (SuitableNumbers[usedCount][i])), usedCount - 1, n);
            }
        }
        return CountByMask[mask];
    }
    
private:
    vector<vector<int>> SuitableNumbers;
    unsigned long long CountByMask[1 << 15];
    bool WasCalced[1 << 15];  
};
__________________________________________________________________________________________________
sample 8136 kb submission
class Solution {
public:
    int ans = 0;
    int n;
    bool used[20];
    void dfs(int x, int y) {
        if(y == n) {
            ans++;
            return;
        }
        used[x] = true;
        for(int i = 1 ; i <= n ; i++) {
            if(!used[i]) {
                if(i % (y+1) == 0 || (y+1) % i == 0) {
                    dfs(i, y+1);
                }
            }
        }
        used[x] = false;
    }
    
    int countArrangement(int N) {
        memset(used, 0, sizeof(used));
        n = N;
        for(int i = 1 ; i <= N ; i++) {
            dfs(i, 1);
        }
        return ans;
    }
};
__________________________________________________________________________________________________

File: Java/526.java
__________________________________________________________________________________________________
sample 0 ms submission
class Solution {
    public int countArrangement(int N) {
        int[] retVal = new int[] {
            0,1,2,3,8,10,36,41,132,250,700,750,4010,4237,10680,24679
        };
        return retVal[N];
    }
}
__________________________________________________________________________________________________
sample 31512 kb submission
//本质是permutation，只是在加入下次dfs时，需满足 条件 i%pos==0 || pos%i==0
class Solution {
    int res;
    public int countArrangement(int N) {
        res = 0;
        int[] visited = new int[N+1];
        dfs(N,1,visited);
        return res;
    }
    public void dfs(int N, int pos,int[] visited){
        if(pos > N) {
            res++;
        }
        for(int i = 1; i<=N; i++){
            if((visited[i] == 0) && (pos%i==0 || i%pos ==0)){
                visited[i] = 1;
                dfs(N, pos+1,visited);
                visited[i] = 0;
            }
        }
    }
}
__________________________________________________________________________________________________
